---
title: "Formatieve opdracht RNA-seq"
author: "Omar Sprockel & Ghislaine van Gilse"
date: "20-10-2022"
output: html_document
---

```{r setup, include=FALSE}
#download eventueel de packages als dit nog niet is gebeurd
library(Rsubread)
library(tidyverse)
library(ggplot2)
library(DESeq2)
library(pheatmap)
library("org.Hs.eg.db")
library(GO.db)
library(GOstats)
```

Onderdeel 1a
Korte introductie
Tijdens dit experiment wordt gekeken naar het verschil in genexpressie tussen fibroblasten en pluripotente stamcellen (iPSC). Om dit te onderzoeken worden fibroblasten teruggedraaid naar iPSC. Voor dit experiment worden fibroblasten uit twee proefpersonen geïsoleerd. Vier transcriptiefactoren in deze fibroblasten worden tot overexpressie gebracht (MYC, OCT3/4, SOX2, en KL4) om deze differentiatie terug te draaien, zodat de cel weer een stamcel wordt. De genexpressie van de celtypen is vervolgens gemeten door middel van RNA-sequencing. De data hiervan wordt in onderstaand onderzoek uitgewerkt met behulp van R.

Onderdeel 1b
```{r Download files, eval=FALSE}
conda create -n rnaseq sra-tools
conda activate rnaseq
cd /home/daur2/rnaseq/rnaseq_ipsc
for id in $(cat ipsc_sampledata.csv)
do 
fastq-dump --split-3 --outdir '/home/daur2/rnaseq/rnaseq_ipsc/fastq/.' --gzip $id
done
```

Onderdeel 1c
```{r Check quality, eval=FALSE}
conda install -c bioconda fastqc
fastqc --outdir /home/daur2/rnaseq/rnaseq_ipsc/fastqc_output/ /home/daur2/rnaseq/rnaseq_ipsc/fastq/*.fastq.gz
```
Over het algemeen ziet de kwaliteit van de fastq bestanden er goed genoeg uit om mee te werken. Zowel de per base sequence quality als de per sequence quality scores zien er goed uit.

Onderdeel 1d
```{r alignment, eval=FALSE}
# Download the reference genome using wget
wget -P '/home/daur2/rnaseq/hg38_genome/' https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/GRCh38.primary_assembly.genome.fa.gz

# Unzip the reference genome
gunzip /home/daur2/rnaseq/hg38_genome/*.fa.gz

# Create variable with location of reference genome
hg38_ref <- "/home/daur2/rnaseq/hg38_genome/GRCh38.primary_assembly.genome.fa"

# Create variable with basename of index files
hg38_index <- "/home/daur2/rnaseq/hg38_index/hg38_index"

# Build index (default paramaters)
buildindex(basename = hg38_index,
           reference = hg38_ref,
           gappedIndex = FALSE,
           indexSplit = FALSE)

# Create object with location of fastq files
fastq_dir <- “/home/daur2/rnaseq/rnaseq_ipsc/fastq/”.

# Create object with basename of index files
hg38_index <- "/home/daur2/rnaseq/hg38_index/hg38_index"

# Create object with output dir for bam files
bam_dir <- "/home/daur2/rnaseq/rnaseq_ipsc/bam/"

# Create a vector with sample numbers
samples <- list.files(fastq_dir, pattern = "_[12].fastq.gz") %>%
  str_remove(pattern = "_[12].fastq.gz") %>% unique()

# Align the reads in the fastq files
alignment_statistics <- align(index = hg38_index,
                              readfile1 = paste0(fastq_dir, samples, "_1.fastq.gz"),
                              readfile2 = paste0(fastq_dir, samples, "_2.fastq.gz"),
                              type = "rna",
                              input_format = "gzFASTQ",
                              output_format = "BAM",
                              output_file = paste0(bam_dir, samples, ".bam"),
                              unique = TRUE,
                              nthreads = 10)

# Store the alignment statistics as R object
saveRDS(alignment_statistics, file = paste0(bam_dir, "alignment_statistics.rds"))
```

Onderdeel 1e
```{r grouped_bar_graph, echo = FALSE}
# Read the file into R
alignment_stats <- readRDS("/home/daur2/rnaseq/rnaseq_ipsc/bam/alignment_statistics.rds")

# Calculate the percentage of uniquely mapping fragments
alignment_stats_t <- alignment_stats %>% 
  t %>% 
  as_tibble() %>% 
  mutate(bamfile=colnames(alignment_stats)) %>%
  mutate(perc_unique_mapped = Uniquely_mapped_fragments/Total_fragments*100)

# Plot the percentages in a bar graph
alignment_stats_t %>%
  ggplot(aes(x = bamfile, y = perc_unique_mapped, fill=bamfile)) +
  geom_col() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  coord_cartesian (ylim =c(0,100))+
  labs(title = "Proportion of uniquely mapped fragments for each sample",
       x= "RNA-seq sample",
       y= "Percentage of uniquely mapping fragments")+
  theme(legend.position = "none")
```

Onderdeel 2a
```{r count_table, eval = FALSE}
# Create object with input dir for bam files
bam_dir <- "/home/daur2/rnaseq/rnaseq_ipsc/bam/"

# Create object with output dir for count tables
counts_dir <- "/home/daur2/rnaseq/rnaseq_ipsc/counts/"

# Create vector with names of bam files
bam_files <- list.files(bam_dir, pattern = ".*\\.bam$", full.names = TRUE)

# Count the reads per gene using the in-built NCBI RefSeq annotations
read_counts <- featureCounts(files = bam_files,
                             annot.inbuilt = "hg38",
                             useMetaFeatures = TRUE,
                             strandSpecific = 1,
                             isPairedEnd = TRUE,
                             countReadPairs = TRUE,
                             nthreads = 10)
```

Onderdeel 2b
```{r count_table_bar_graph, echo = FALSE}
# Obtain the dataframe with the statistics
read_counts <- readRDS("/home/daur2/rnaseq/rnaseq_ipsc/counts/read_counts.rds")
count_stats <- read_counts$stat

# Convert first column to rownames
rownames(count_stats) <- count_stats$Status
count_stats$Status <- NULL

# Calculate the percentage of uniquely counting fragments
count_stats_t <- count_stats %>% 
  t %>% 
  as_tibble() %>% 
  mutate(bamfile=colnames(count_stats)) %>%
  mutate(Total=colSums(count_stats)) %>%
  mutate(perc_assigned = Assigned/Total*100)

# Plot the percentages in a bar graph
count_stats_t %>% ggplot(aes(x = bamfile, y = perc_assigned, fill=bamfile)) +
  geom_col() +
  ggtitle("Proportion of counted fragments for each sample") +
  xlab("RNA-seq sample") +
  ylab("Percentage of counted fragments") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90), legend.position = "none") +
  coord_cartesian (ylim =c(0,100))
```


Onderdeel 2c
```{r make DESeq2 object, echo = FALSE}
read_counts <- readRDS("/home/daur2/rnaseq/rnaseq_ipsc/counts/read_counts.rds")

# Obtain the count matrix
count_matrix <- read_counts$counts

# Import the sample data
metadata <- read_csv("/home/daur2/rnaseq/rnaseq_ipsc/ipsc_sampledata.csv")

# Convert the metadata to dataframe object
metadata <- as.data.frame(metadata)

# Add rownames to the metadata dataframe
rownames(metadata) <- paste0(metadata$Run, ".bam")

# Create a column specifying the test condition
metadata$Cell_type <- metadata$Cell_type %>% factor(levels = c("Skin derived fibroblast", "iPSC"))

# Create the DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = count_matrix, colData = metadata, design = ~ Cell_type)
```

Onderdeel 2d
```{r PCA analyse, echo=FALSE}
dds_normalized <- rlog(dds)

# Perform PCA on the ipsc study dataset
pca <- dds_normalized %>% assay() %>% t() %>% prcomp()

# Find the proportion of variance explained by each PC
pca_summary <- summary(pca)$importance
pca_summary

# Create data frame with metadata and the PCA values for plotting
pca_plotting <- cbind(metadata, pca$x)

# Obtain the percentages of variation covered by PC1 and PC2
PC1_var <- round(pca_summary["Proportion of Variance", "PC1"]*100, digits = 1)
PC2_var <- round(pca_summary["Proportion of Variance", "PC2"]*100, digits = 1)

# Plot PC1 vs PC2
ggplot(pca_plotting) + 
  geom_point(aes(x=PC1, y=PC2, color = source_name, shape = source_name), size = 5) +
  ggtitle("PCA for iPSC study") +
  xlab(paste0("PC1 (", PC1_var, "%)")) +
  ylab(paste0("PC2 (", PC2_var, "%)")) +
  theme_bw()+
  labs(color="Cell type", shape="Cell type")
```

Onderdeel 2e
```{r Heatmap, echo = FALSE}
# Extract the matrix from the normalized DESeq object
dds_normalized_matrix <- assay(dds_normalized)    

# Compute pairwise correlation values
ipsc_cor <- cor(dds_normalized_matrix)    
ipsc_cor

# Plot heatmap using the correlation; annotate the heatmap based on the treatment
pheatmap(ipsc_cor, annotation = metadata["source_name"])
```

Onderdeel 3a
```{r DGEanalysis, echo = FALSE, message=FALSE}
# Perform the DGE analysis using DESeq2
ipsc_dge <- DESeq(dds)
# Obtain the results for the DGE analysis
ipsc_dge_results <- results(ipsc_dge)
ipsc_dge_results
summary(ipsc_dge_results)
```

Onderdeel 3b
```{r Volcano plot, echo = FALSE}
# Obtain the genes with a significant p-value
sign_genes <- ipsc_dge_results[which(ipsc_dge_results$padj < 0.05),]

# Obtain the id of the most significantly upregulated gene
topGene <- sign_genes[which.max(sign_genes$log2FoldChange),]
topGene_name <- rownames(topGene)

# Create a dataframe for plotting without genes with padj = NA
ipsc_dge_plotting <- data.frame(ipsc_dge_results) %>% filter(!is.na(padj))

# Create column specifying if gene is significantly differentially expressed
ipsc_dge_plotting <- ipsc_dge_plotting %>% 
  mutate(signif = if_else(padj < 0.05 & abs(log2FoldChange) >1, "Significant", "Not significant"))

# Create a volcano plot
ipsc_dge_plotting %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = signif)) +
  geom_point() + 
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") + 
  theme_bw() +
  
  # Change the legend text and the colours
  scale_colour_manual(values = c("grey", "darkblue"), name = "Significance") +
  
  # Annotate the most significantly upregulated gene
  annotate("text", x = topGene$log2FoldChange, y = -log10(topGene$padj)*0.8, 
           label = topGene_name, colour = "blue")+

# Indicate thresholds as lines
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  geom_vline(xintercept = 1, linetype = "dashed") +
  geom_vline(xintercept = -1, linetype = "dashed")

```

Onderdeel 3c
```{r HeatMap, echo = FALSE}
# Obtain the gene identifiers of the 15 most significant genes
top15_genes <- rownames(ipsc_dge_results[order(ipsc_dge_results$padj)[1:15],])

# Obtain the count values for these genes
count_values <- assay(dds)[top15_genes,]

# Replace the column names with conditions
colnames(count_values) <- colData(dds)$Cell_type

# Visualize the results in a heatmap
pheatmap(count_values, scale = "row", show_rownames = TRUE)
```

Onderdeel 4a
```{r HeatMap_gene_symbol, eval = FALSE}
# Obtain the gene identifiers of the 15 most significant genes
top15_genes <- rownames(ipsc_dge_results[order(ipsc_dge_results$padj)[1:15],])

# Obtain the count values for these genes
count_values <- assay(dds)[top15_genes,]

# Replace the column names with conditions
colnames(count_values) <- colData(dds)$Cell_type

# Visualize the results in a heatmap
pheatmap(count_values, show_rownames = TRUE)
```

Onderdeel 4b
```{r GO_term_analyse_functie, eval = FALSE}

GO_terms <- function(a, b, c, d) {
DESeq2 <- a[,]
Enrichment_analyse <- b[,]
LFC <- c[,]
pwaarde <- d[,]

if (Enrichtment_analyse == “TRUE”)
{upregulated_genes <- DESeq2 %>% data.frame() %>% 
  filter(log2FoldChange > LFC, padj pwaarde) %>% rownames()

# Create a list of all genes in the dataset
all_genes <- ipsc_dge_results %>% data.frame() %>% rownames()

# Perform GO term enrichment analysis
test_object <- new("GOHyperGParams",
                   geneIds = upregulated_genes,
                   universeGeneIds = all_genes, 
                   annotation = "org.Hs.eg.db", 
                   ontology = "BP", 
                   pvalueCutoff = 1,
                   testDirection = "over")
goterm_analysis <- hyperGTest(test_object)
return(summary(goterm_analysis))

} if else (Enrichtment_analyse == “FALSE”){
{downregulated_genes <- DESeq2 %>% data.frame() %>% 
  filter(log2FoldChange > LFC, padj pwaarde) %>% rownames()

# Create a list of all genes in the dataset
all_genes <- ipsc_dge_results %>% data.frame() %>% rownames()

# Perform GO term enrichment analysis
test_object <- new("GOHyperGParams",
                   geneIds = upregulated_genes,
                   universeGeneIds = all_genes, 
                   annotation = "org.Hs.eg.db", 
                   ontology = "BP", 
                   pvalueCutoff = 1,
                   testDirection = "over")
goterm_analysis <- hyperGTest(test_object)
return(summary(goterm_analysis))

a	DESeq2 results object (results(DESeq(dds)))
b	Logische waarde (TRUE of FALSE) die aangeeft of de enrichment analyse gedaan moet worden voor de upgereguleerde genen of voor de downgereguleerde genen.
c	Drempelwaarde voor |LFC|.
d	Drempelwaarde voor adjusted p-waarde.

Bijvoorbeeld: gotermAnalysis(ipsc_dge_results, upregulated = FALSE, 1, 0.01) geeft de summary van de enrichment analyse voor de downgereguleerde genen met een LFC < -1 en een adjusted p-waarde < 0.01 voor de genen in het DESeq2 results object ipsc_dge_results.
```

Onderdeel 4c
```{r GO_terms_graph, echo = FALSE}
# Create a list of upregulated genes
upregulated_genes <- ipsc_dge_results %>% data.frame() %>% 
  filter(log2FoldChange > 1, padj < 0.01) %>% rownames()

# Create a list of all genes in the dataset
all_genes <- ipsc_dge_results %>% data.frame() %>% rownames()

# Perform GO term enrichment analysis
test_object <- new("GOHyperGParams",
                   geneIds = upregulated_genes,
                   universeGeneIds = all_genes, 
                   annotation = "org.Hs.eg.db", 
                   ontology = "BP", 
                   pvalueCutoff = 1,
                   testDirection = "over")
goterm_analysis <- hyperGTest(test_object)

# Obtains dataframe with results of GO term analysis
goterm_analysis_results <- summary(goterm_analysis)

# Adjust the p values for multiple testing
goterm_analysis_results$padj <- p.adjust(goterm_analysis_results$Pvalue, method = "BH")

# Select only gene sets that are larger than 5 but smaller 
# than 500 (to prevent taking along very small and very large
# gene sets).
goterm_analysis_results <- goterm_analysis_results %>% filter(Count > 5) %>% filter(Count < 500)

# Select the top 20 GO terms
goterm_analysis_top20 <- goterm_analysis_results[order(goterm_analysis_results$padj)[1:20],]

# Plot the p-values of the top 20 GO terms
goterm_analysis_top20$Term <- factor(goterm_analysis_top20$Term, 
                                     levels = goterm_analysis_top20$Term[
                                       order(goterm_analysis_top20$padj, decreasing = TRUE)])
goterm_analysis_top20 %>% ggplot(aes(x = Term, y = -log10(padj))) +
  geom_point() +
  coord_flip() +
  ylab(expression(-log[10](adjusted~italic(P)~value))) + 
  xlab("GO terms") +
  ggtitle("Top 20 enriched GO terms\n for upregulated genes") +
  theme_bw()



# Create a list of downregulated genes
downregulated_genes <- ipsc_dge_results %>% data.frame() %>% 
  filter(log2FoldChange < -1, padj < 0.01) %>% rownames()

# Create a list of all genes in the dataset
all_genes2 <- ipsc_dge_results %>% data.frame() %>% rownames()

# Perform GO term enrichment analysis
test_objec2t <- new("GOHyperGParams",
                   geneIds = downregulated_genes,
                   universeGeneIds = all_genes, 
                   annotation = "org.Hs.eg.db", 
                   ontology = "BP", 
                   pvalueCutoff = 1,
                   testDirection = "over")
goterm_analysis2 <- hyperGTest(test_object)

# Obtains dataframe with results of GO term analysis
goterm_analysis_results2 <- summary(goterm_analysis2)

# Adjust the p values for multiple testing
goterm_analysis_results2$padj <- p.adjust(goterm_analysis_results2$Pvalue, method = "BH")

# Select only gene sets that are larger than 5 but smaller 
# than 500 (to prevent taking along very small and very large
# gene sets).
goterm_analysis_results2 <- goterm_analysis_results2 %>% filter(Count > 5) %>% filter(Count < 500)

# Select the top 20 GO terms
goterm_analysis2_top20 <- goterm_analysis_results2[order(goterm_analysis_results2$padj)[1:20],]

# Plot the p-values of the top 20 GO terms
goterm_analysis2_top20$Term <- factor(goterm_analysis2_top20$Term, 
                                     levels = goterm_analysis2_top20$Term[
                                       order(goterm_analysis2_top20$padj, decreasing = TRUE)])
goterm_analysis2_top20 %>% ggplot(aes(x = Term, y = -log10(padj))) +
  geom_point() +
  coord_flip() +
  ylab(expression(-log[10](adjusted~italic(P)~value))) + 
  xlab("GO terms") +
  ggtitle("Top 20 enriched GO terms\n for downregulated genes") +
  theme_bw()
```
