---
title: "daur_les3"
author: "Ghislaine van Gilse"
date: "10/19/2022"
output: html_document
---

```{r setup, include=FALSE}

library(Rsubread)
library(tidyverse)
library(ggplot2)
library(DESeq2)
library(pheatmap)

```

Exercise 3.1
Suppose we find a gene in the airway study that has an adjusted p-value < 0.05 and shows a LFC of -1.45 when treated cells are compared to untreated cells. How would you interpret this result?
2^1.45 = 2.7 = 3x, decreasing. Aka downregulation.
This gene is significantly downregulated in response to the treatment. The LFC corresponds to a downregulation of 21.45 x, which equals 3x.

```{r DGE analysis}
# Perform the DGE analysis using DESeq2
airway_dge <- DESeq(dds)

# Obtain the results for the DGE analysis
airway_dge_results <- results(airway_dge)
airway_dge_results

# Obtain the summary of the results
summary(airway_dge_results)

```

Exercise 3.2
We would like to change the default settings of the results() function. Instead of reporting genes with a LFC bigger or smaller than 0, we would like to report genes with a LFC < -1 (downregulated genes) or LFC > 1 (upregulated genes). And instead of reporting genes with an adjusted p-value < 0.1, we would like to report the genes with an adjusted p-value < 0.05.

a) What does a LFC of 1 mean? How much did the gene expression then actually change?
  Een LFC van 1 betekend upregulatie 1x, dus 2^1 = 2.
  A LFC > 1 means that the gene is upregulated 2^1 times, so 2 times. Similarly, a LFC < -1 means that the gene is downregulated 2 times.

b) Look into the documentation of the results() function to find out which default options have to be changed. Change these options to create new results.
  # Obtain the results for the DGE analysis
airway_dge_results2 <- results(airway_dge, alpha = 0.05, lfcThreshold = 1)

  # Obtain the summary of the results
summary <- summary(airway_dge_results2)

How many genes are now considered to be upregulated? And how many genes are downregulated?
  68 up and 36 down


# Obtain the genes with a significant p-value
sign_genes <- airway_dge_results[which(airway_dge_results$padj < 0.05),]

# Obtain the id of the most significantly upregulated gene
topGene <- sign_genes[which.max(sign_genes$log2FoldChange),]
topGene_name <- rownames(topGene)
topGene_name

```{r plotCounts}
# Obtain the gene counts using the plotCounts function
geneCounts <- plotCounts(dds, gene = topGene_name, 
                         intgroup = c("treatment"), 
                         returnData = TRUE)

# Create a counts plot using ggplot
ggplot(geneCounts, aes(x = treatment, y = count)) +
  scale_y_log10() + 
  geom_point(position = position_jitter(width = 0.1, height = 0), 
             size = 3, colour = "darkgreen") +
  xlab("Dexamethasone treatment") +
  ylab("Fragment count") + 
  ggtitle(topGene_name) +
  theme_bw()
```

Exercise 3
Create a count plot for the most downregulated gene.

Hint: use the which.min() function.
```{r whichmin}
# Obtain the id of the most significantly downregulated gene
topGene_down <- sign_genes[which.min(sign_genes$log2FoldChange),]
topGene_down_name <- rownames(topGene_down)

# Obtain the gene counts using the plotCounts function
geneCounts <- plotCounts(dds, gene = topGene_down_name, 
                         intgroup = c("treatment"), 
                         returnData = TRUE)

# Create a counts plot using ggplot
ggplot(geneCounts, aes(x = treatment, y = count)) +
  scale_y_log10() + 
  geom_point(position = position_jitter(width = 0.1, height = 0), 
             size = 3, colour = "darkgreen") +
  xlab("Dexamethasone treatment") +
  ylab("Fragment count") + 
  ggtitle(topGene_down_name) +
  theme_bw()
```

```{r geen NA}
# Create a dataframe for plotting without genes with padj = NA
airway_dge_plotting <- data.frame(airway_dge_results) %>% filter(!is.na(padj))

# Create column specifying if gene is significantly differentially expressed
airway_dge_plotting <- airway_dge_plotting %>% 
  mutate(signif = if_else(padj < 0.05, "padj < 0.05", "Not significant"))

# Create a volcano plot
airway_dge_plotting %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = signif)) +
  geom_point() + 
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") + 
  theme_bw() +
  
  # Change the legend text and the colours
  scale_colour_manual(values = c("grey", "darkgreen"), name = "Significance") +
  
  # Annotate the most significantly upregulated gene
  annotate("text", x = topGene$log2FoldChange, y = -log10(topGene$padj)*0.8, 
           label = topGene_name, colour = "blue")
```

Exercise 3
Create another Volcano plot, in which you colour all genes with an adjusted p-value < 0.01 and  
|LFC|>1 darkred. Also indicate the LFC and p-value thresholds as dashed lines in the plot.

Hint: you can plot horizontal and vertical lines using the functions geom_hline() and geom_vline(), respectively.

```{r Volcano plot 2}
# Create a dataframe for plotting without genes with padj = NA
airway_dge_plotting <- data.frame(airway_dge_results) %>% filter(!is.na(padj))

# Create column specifying if gene is significantly differentially expressed
airway_dge_plotting <- airway_dge_plotting %>% 
  mutate(signif = if_else(padj < 0.01 & abs(log2FoldChange) > 1, 
                          "Significant", "Not significant"))

# Create a volcano plot
airway_dge_plotting %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = signif)) +
  geom_point() + 
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") + 
  theme_bw() +
  
  # Change the legend text and the colours
  scale_colour_manual(values = c("grey", "darkred"), name = "Significance") +

  # Indicate thresholds as lines
  geom_hline(yintercept = -log10(0.01), linetype = "dashed") +
  geom_vline(xintercept = 1, linetype = "dashed") +
  geom_vline(xintercept = -1, linetype = "dashed")
```

```{r heatmap1}
# Obtain the gene identifiers of the 10 most significant genes
top10_genes <- rownames(airway_dge_results[order(airway_dge_results$padj)[1:10],])

# Obtain the count values for these genes
count_values <- assay(dds)[top10_genes,]

# Replace the column names with conditions
colnames(count_values) <- colData(dds)$treatment

# Visualize the results in a heatmap
pheatmap(count_values, show_rownames = TRUE)
```

```{r heatmap2}
# Visualize the results in a heatmap (with scaling!)
pheatmap(count_values, scale = "row", show_rownames = TRUE)
```

Exercise 3.5
How many of the top 10 most significantly DE genes are upregulated by the treatment? And how many genes are downregulated by the treatment?
  9 uit de 10. Alleen de bovenste rij gaat van rood naar blauw, die is downregulated.

Could you also determine this based on the unscaled heatmap?
  This is also visible in the unscaled heatmap, but it is less clear. The scaled heatmap is more easy to interpret when it comes to comparing the different samples with each other. (5e rij enige van licht naar donker, hoog naar laag dus).

