---
title: "daur2"
author: "Ghislaine van Gilse"
date: "10/17/2022"
output: html_document
---

```{r setup}

library(Rsubread)
library(tidyverse)
library(ggplot2)

```


Exercise 1.1
Go to the GEO page of the airway study. Try to answer the following questions with the information on the GEO website:

a. What type of ASM cells were used in the experiment?
  HASM(Human airway smooth muscle) cells derived from four different donors.
b. Which glucocorticoid was used in the experiments and how long were the cells exposed to this glucocorticoid?
  Treatment with a glucocorticosteroid (i.e. Dexamethasone (Dex), 1Î¼M for 18h)
c. Which sequencing machine was used to generate the RNA-seq data of the airway study?
  an Illumina Hi-Seq 2000 instrument.
d. Is the RNA-seq data of the airway study single-end or paired-end sequencing data?
  Paired-end sequencing data

Exercise 1.2
fastq-dump --split-3 --outdir '/home/bas.vangestel/daur2/rnaseq_airway/fastq/.' --gzip SRR1039508

Use the help file of the fastq-dump tool (fastq-dump --help) to find out what the different options in the command above do.
--split-3   Split spot. Plek waar je een splitsing zet waarbij ze worden opgeslagen in aparte files. Elke file bevat dan gelijke                    blokken van 3 zegmaar
--outdir    Output directory. Plek waar je je output zet
--gzip      Compress output using gzip. Aka zipfile maken

Exercise 1.3
Explain why each sample/SRA identifier has two or sometimes three fastq files.
  The sequencing data is paired-end sequencing data. The forward and reverse read of the sequencing data are stored in separate fastq     files. If one of the two reads is missing (only forward or only reverse read is available), the read is stored in a separate,         third file.
  
Exercise 1.4
Write a for loop in bash to run the fastq-dump for all the SRA identifiers.

#```{r for loop}
for id in $(cat sraid.txt)
do 
fastq-dump --split-3 --outdir '/home/bas.vangestel/daur2/rnaseq_airway/fastq/.' --gzip $id
done
#```

conda install -c bioconda fastqc

# Perform fastqc analysis
fastqc --outdir ./rnaseq_airway/fastqc_output/ ./rnaseq_airway/fastq/*.fastq.gz


Exercise 1.5
Check the html reports that were generated by the fastqc tool. What are your conclusions?
  In general, the quality of the fastq files is good. This is true for both the per base sequence quality and for the per sequence     quality scores. We can therefore use the fastq files for downstream analysis.

#To activate a conda:
  conda activate myenv

#To deactivate a conda:
  conda deactivate #(gaat automoatisch terug naar base)

Exercise 1.6
Check the help page for the buildindex function. What do the options gappedIndex = FALSE and indexSplit = FALSE mean?
  gappedIndex = F   Indicates that a full genome index has to be build. The alternative is a gapped index, which is smaller                       in size (takes less space on your computer), but is less efficient when it comes to mapping your reads to                     the reference genome.
  indexSplit = F    Indicates that the genome index file should be stored as one big file and not as several smaller files.


Exercise 1.7
Check the help page for the align function. What does the options unique = TRUE mean? Why would it be set to TRUE?
  unique = T    The option indicates that only uniquely mapping reads should be reported. Some reads align to multiple locations in the genome (and do this with the same alignment score). For these reads, it is difficult to determine from which transcript (and hence from which gene) the reads were generated. It is therefore better to only include reads in the analysis that align to only one location in the genome; these reads are the uniquely mapping reads.

```{r reads}
# Read the file into R
alignment_stats <- readRDS("/home/daur2/rnaseq/rnaseq_airway/bam/alignment_statistics.rds")

# Calculate the percentage of uniquely mapping fragments
alignment_stats_t <- alignment_stats %>% 
  t %>% 
  as_tibble() %>% 
  mutate(bamfile=colnames(alignment_stats)) %>%
  mutate(perc_unique_mapped = Uniquely_mapped_fragments/Total_fragments*100)

# Plot the percentages in a bar graph
alignment_stats_t %>%
  ggplot(aes(x=bamfile, y=perc_unique_mapped, fill=bamfile)) +
  geom_col()+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90)) +
  coord_cartesian (ylim =c(0,100))+
  labs(title = "Proportion of uniquely mapped fragments for each sample",
       x= "RNA-seq sample",
       y= "Percentage of uniquely mapping fragments")+
  theme(legend.position = "none")
```

Conclusion: Based on this graph you can conclude that about 90% of the fragments uniquely maps to the human genome.